crates/agent/src/publications.rs

```diff
@@ -196,9 +201,54 @@ impl PublishHandler {
             .await;
         }
 
-        let live_spec_ids: Vec<_> = spec_rows.iter().map(|row| row.live_spec_id).collect();
+        let expanded_rows = specs::expanded_specifications(row.user_id, &spec_rows, txn).await?;
+        tracing::debug!(specs = %expanded_rows.len(), "resolved expanded specifications");
+
+        let errors = draft::extend_catalog(
+            &mut draft_catalog,
+            expanded_rows
+                .iter()
+                .map(|r| (r.live_type, r.catalog_name.as_str(), r.live_spec.0.as_ref())),
+        );
+        if !errors.is_empty() {
+            anyhow::bail!("unexpected errors from expanded specs: {errors:?}");
+        }
+
+        // It's important that we prune unbound collection specs _after_ the catalog is expanded,
+        // and before checking resource quotas
+        let pruned_collections =
+            specs::prune_unbound_collections(&mut draft_catalog, &mut spec_rows);
+        // `resolve_specifications` will have created `live_specs` for each of the `draft_specs` in the publication.
```

Comment:
Interesting... writing to `live_specs` ought to be one of the last steps of publishing, no? Does this mean that we temporarily leak incorrect `live_specs`? Does that not matter because everything here is happening inside a transaction?

Reply:
Yeah, we're inside a transaction, so nothing else will observe these changes unless it commits. We do write out the new live specs fairly early during the publication, as it simplifies some things. The new approach for pruning unbound collections will actually update all the live specs and then delete the ones that are unbound.